### DISCUSSION TITLE
Some ideas for the library

### DISCUSSION THREAD

**Discussion by duongkimlam31** [Disservin/chess-library#56] [Ideas]
1. Generate legal captures
2. Generate legal non-captures
3. Check if a move is legal
4. Check if FEN is legal
5. Show the location of the attackers from a given square

**Comment by Disservin** [comment]
1. and 2. are already possible, see https://disservin.github.io/chess-library/pages/Move-Generation.html#move-generation
specifically `MoveGenType::ALL, MoveGenType::CAPTURE, MoveGenType::QUIET`.

3. I will think about a fast alternative but the current way would be to generate all legal moves and use .find on the movelist to check if the move exists.
4. Yeah not implemented, but a todo
5. How do you imagine `show` ? I could add such a function which just returns a bitboard for example.

**Reply by duongkimlam31** [reply]
For #4, maybe return the attacker squares. That way, the users can manipulate those squares such as using them to determine the piece types. Also, you guys are doing great. Keep up the work!

**Reply by duongkimlam31** [reply]
There should also be an option to get the ply of the board as it is important for techniques such as killer move heuristic and mate-distance pruning.

**Reply by Disservin** [reply]
> There should also be an option to get the ply of the board as it is important for techniques such as killer move heuristic and mate-distance pruning.

Yes, but no, the ply would be the boards fullmovenumber essentially multiplied by 2 and taking into account the sidetomove. Maximum chess game length is what? 6000? Idk. Mostly people have a limit to their search like MAX_DEPTH = 200, and there the ply 0 is wherever the search has started and each recursive call adds 1. If you instead use .ply() from the board you dont really know the max and how large your killer move array should be.

**Reply by duongkimlam31** [reply]
I'm using a hash table to store my killer moves array, so having a getPly() function may be beneficial in this case. The problem is that I don't know whether using a hash table provides more benefits than using the array or not. Not only that, having getPly() also helps with other techniques such as mate-distance pruning, not just killer moves heuristic.

**Reply by duongkimlam31** [reply]
Also, can you guys implement a function that allows the users to read opening books? This is what I have so far using the library.
[polyglot.zip](https://github.com/Disservin/chess-library/files/13868097/polyglot.zip). The code is based on Bluefever software.

**Reply by Disservin** [reply]
> I'm using a hash table to store my killer moves array, so having a getPly() function may be beneficial in this case. The problem is that I don't know whether using a hash table provides more benefits than using the array or not. Not only that, having getPly() also helps with other techniques such as mate-distance pruning, not just killer moves heuristic.

using the ply of the board position is conceptually wrong for these things. You really shouldnt use a hash table either (definitely not std::map or std::unordered_map). Just use a fixed size array and limit your search depth. 
I can think about adding polyglot support, reading pgns is already supported btw.

**Reply by duongkimlam31** [reply]
Ah, I see. I thought the ply was the distance from the initial position not the distance from the root. Thank you for pointing that out.

**Reply by duongkimlam31** [reply]
I'm currently using [https://github.com/dshawul/nnue-probe](url) to develop an nnue evaluation function for my engine. This is the description of the function:
/*
 Evaluation subroutine suitable for chess engines.
 Piece codes are
     wking=1, wqueen=2, wrook=3, wbishop= 4, wknight= 5, wpawn= 6,
     bking=7, bqueen=8, brook=9, bbishop=10, bknight=11, bpawn=12,
 Squares are
     A1=0, B1=1 ... H8=63
 Input format:
     piece[0] is white king, square[0] is its location
     piece[1] is black king, square[1] is its location
     ..
     piece[x], square[x] can be in any order
     ..
     piece[n+1] is set to 0 to represent end of array
 Returns
   Score relative to side to move in approximate centi-pawns

*/

DLLExport int _CDECL nnue_evaluate(
  int player,                       /** Side to move: white=0 black=1 */
  int* pieces,                      /** Array of pieces */
  int* squares                      /** Corresponding array of squares each piece stands on */
); 

My question is whether it is possible to get the pieces and their corresponding squares using the current library.

**Reply by Disservin** [reply]
Squares have the same encoding, on square you can just call `.index()`, pieces have different encodings. So you will have to write some manual conversion function.

**Reply by duongkimlam31** [reply]
I'm thinking of making a function that takes in a chess::Board object. From that, I want to know all the available pieces and their corresponding squares. Does this sound feasible?

**Reply by Disservin** [reply]
Sure why wouldnt it ? Btw have you taken a look at https://disservin.github.io/chess-library/pages/board.html?

**Reply by duongkimlam31** [reply]
I have looked at the link you provided but I can't use it to my advantage for a couple of reasons:
1. Not all the functions in the library are available on the website.
2. The documentation doesn't provide all the information needed to make it easier to understand and use.

I don't blame it as the library is enormous and it is still under development which makes documentation rather a daunting task.

**Reply by duongkimlam31** [reply]
For example, I want to get all the available pieces on the board for my nnue evaluation function, which function should I use? From the link provided, I see two functions that stand out to me, which are pieces with different signatures. However, due to the insufficient documentation, I don't know how to use it.

**Reply by Disservin** [reply]
I see, well, you can use occ(), which returns a bitboard with all bits set where a piece is then you loop over all set bits
```cpp
auto occ = board.occ();

while (occ) {
  auto sq = occ.pop();
  auto piece = board.at(sq);
}
```

Pieces also returns a bitboard with all bits set but only for the specified parameters, i.e. only passing the type will get you white and black piece types locs and passing the color will only get piecetypes of that color

**Reply by duongkimlam31** [reply]
I'm currently using version 0.5.13. Is there an equivalent of pop() in this version? It is poplsb() correct?

**Reply by Disservin** [reply]
Yes that’s correct

**Comment by Hyvok** [comment]
I'd also really like to have the ability to check if a move is legal. Other than that library seems very nice for my use case, nice work!

**Comment by Razamindset** [comment]
Awesome library. can't complain anything. But the web docs can be improved a little bit
